File: src/Java/myservicebus/src/main/java/com/myservicebus/persistence/OutboxSession.java
package com.myservicebus.persistence;

import java.util.Objects;

public final class OutboxSession {
    private OutboxWriter writer;

    /**
     * Activates a transactional writer for scoped publish and send endpoints until the registration is closed.
     *
     * @throws IllegalStateException when an outbox transaction is already active in this service scope
     */
    public synchronized Registration begin(OutboxWriter outboxWriter) {
        Objects.requireNonNull(outboxWriter, "outboxWriter");
        if (writer != null) {
            throw new IllegalStateException("An outbox transaction is already active in this service scope.");
        }
        writer = outboxWriter;
        return new Registration(this, outboxWriter);
    }

    public synchronized OutboxWriter getWriter() {
        return writer;
    }

    public static final class Registration implements AutoCloseable {
        private OutboxSession session;
        private final OutboxWriter writer;

        private Registration(OutboxSession session, OutboxWriter writer) {
            this.session = session;
            this.writer = writer;
        }

        @Override
        public synchronized void close() {
            if (session != null) {
                session.clear(writer);
                session = null;
            }
        }
    }

    private synchronized void clear(OutboxWriter expected) {
        if (writer == expected) {
            writer = null;
        }
    }
}

File: src/MyServiceBus/Persistence/OutboxSession.cs
namespace MyServiceBus.Persistence;

public sealed class OutboxSession
{
    private IOutboxWriter? writer;

    internal IOutboxWriter? Writer => Volatile.Read(ref writer);

    /// <summary>
    /// Activates the supplied transactional writer for scoped publish and send endpoints until the returned
    /// registration is disposed.
    /// </summary>
    /// <exception cref="InvalidOperationException">An outbox transaction is already active in this service scope.</exception>
    public IDisposable Begin(IOutboxWriter outboxWriter)
    {
        ArgumentNullException.ThrowIfNull(outboxWriter);
        if (Interlocked.CompareExchange(ref writer, outboxWriter, null) is not null)
            throw new InvalidOperationException("An outbox transaction is already active in this service scope.");
        return new Registration(this, outboxWriter);
    }

    private sealed class Registration : IDisposable
    {
        private OutboxSession? session;
        private readonly IOutboxWriter writer;

        public Registration(OutboxSession session, IOutboxWriter writer)
        {
            this.session = session;
            this.writer = writer;
        }

        public void Dispose()
        {
            var current = Interlocked.Exchange(ref session, null);
            if (current is not null)
                Interlocked.CompareExchange(ref current.writer, null, writer);
        }
    }
}
